sslSniffer/sslSnifferTest/snifftest.c

```diff
@@ -569,6 +580,320 @@ static int SnifferAsyncPollQueue(byte** data, char* err, SSLInfo* sslInfo,
 }
 #endif /* WOLFSSL_ASYNC_CRYPT */
 
+#ifdef THREADED_SNIFFTEST
+
+/* Change to increase number of threads */
+const int workerThreadCount = 3;
+
+typedef struct {
+    volatile int lockCount;
+    pthread_mutex_t mutex;
+    pthread_cond_t cond;
+} wm_Sem;
+
+ /* Posix style semaphore */
+static int wm_SemInit(wm_Sem *s){
+    s->lockCount = 0;
+    pthread_mutex_init(&s->mutex, NULL);
+    pthread_cond_init(&s->cond, NULL);
+    return 0;
+}
+static int wm_SemFree(wm_Sem *s){
+    pthread_mutex_destroy(&s->mutex);
+    pthread_cond_destroy(&s->cond);
+    return 0;
+}
+static int wm_SemLock(wm_Sem *s){
+    pthread_mutex_lock(&s->mutex);
+    while (s->lockCount > 0)
+        pthread_cond_wait(&s->cond, &s->mutex);
+    s->lockCount++;
+    pthread_mutex_unlock(&s->mutex);
+    return 0;
+}
+static int wm_SemUnlock(wm_Sem *s){
+    pthread_mutex_lock(&s->mutex);
+    s->lockCount--;
+    pthread_cond_signal(&s->cond);
+    pthread_mutex_unlock(&s->mutex);
+    return 0;
+}
+
+typedef struct SnifferWorker {
+    SnifferPacket *head; /* head for doubly-linked list of sniffer packtets */
+    SnifferPacket *tail; /* tail for doubly-linked list of sniffer packtets */
+    wm_Sem         sem;
+    pthread_t      tid;
+    char *server;
+    char *keyFilesSrc;
+    char *passwd;
+    int   port;
+    int   hadBadPacket;  /* track if sniffer worker saw bad packet */
+    int   lastProcessed; /* store last processed packet number */
+    int   shutdown;
+} SnifferWorker;
+
+static int ssl_Init_SnifferWorker(SnifferWorker* worker, int port,
+        const char* server, const char* keyFilesSrc, const char* passwd)
+{
+    wm_SemInit(&worker->sem);
+
+    worker->server      = (char*)server;
+    worker->keyFilesSrc = (char*)keyFilesSrc;
+    worker->passwd      = (char*)passwd;
+    worker->port           = port;
+    worker->shutdown       = 0;
+    worker->lastProcessed  = 0;
+
+    worker->head = (SnifferPacket*)XMALLOC(sizeof(SnifferPacket), NULL,
+                           DYNAMIC_TYPE_TMP_BUFFER);
+
+    if (worker->head == NULL) {
+        XFREE(worker->head, NULL, DYNAMIC_TYPE_TMP_BUFFER);
+        return MEMORY_E;
+    }
+
+    XMEMSET(worker->head, 0, sizeof(SnifferPacket));
+
+    worker->tail = worker->head;
+    worker->head->valid = 0;
+
+    return 0;
+}
+
+static void ssl_Free_SnifferWorker(SnifferWorker* worker)
+{
+    wm_SemFree(&worker->sem);
+
+    if (worker->head != NULL) {
+        XFREE(worker->head, NULL, DYNAMIC_TYPE_TMP_BUFFER);
+    }
+}
+
+static int SnifferWorkerPacketAdd(SnifferWorker* worker, int lastRet,
+        void* chain, int chainSz, int isChain, int packetNumber)
+{
+    SnifferPacket* newEntry;
+    byte*          packet;
+    int            length;
+
+#ifdef WOLFSSL_SNIFFER_CHAIN_INPUT
+    if (isChain) {
+        struct iovec* vchain = (struct iovec*)chain;
+        length = 0;
+        for (i = 0; i < chainSz; i++)
```

Comment:
Please test with `WOLFSSL_SNIFFER_CHAIN_INPUT`. `i` is not defined.

Reply:
Fixed and tested with `WOLFSSL_SNIFFER_CHAIN_INPUT`